iOS 13 Context Menu

justinmc opened the pull request:
## Description

This PR implements the `ContextMenu` widget.

```dart
ContextMenu(
  child: FittedBox(
    fit: BoxFit.cover,
    child: Image.asset('images/my_image.jpg'),
  ),
  actions: <ContextMenuSheetAction>[
    ContextMenuSheetAction(
      child: const Text('My action'),
      onPressed: () { ... },
    ),
    // More ContextMenuSheetActions...
  ],
);
```

| native | Flutter |
| -------- | -------- |
| <img width="140" src="https://user-images.githubusercontent.com/389558/62391329-549e9680-b519-11e9-8535-e84e3af7837d.gif" /> | ![context_menu_small](https://user-images.githubusercontent.com/389558/64367640-65857080-cfcd-11e9-8314-d2912a4a7c40.gif) |


## Related Issues

Closes https://github.com/flutter/flutter/issues/34728

## Tests

  * ContextMenuSheetAction
    - tapping
    - hold to change color
  * ContextMenu
    - default state without being interacted with
    - opening
    - all different ways to close

justinmc committed:
ContextMenu widget that just does the bounce longpress

justinmc committed:
Possible way to do add background blur to existing modal code

justinmc committed:
An option for adding widget to popup as an image

justinmc committed:
Children in and out of popup line up perfectly

justinmc committed:
Animate child to final open position, hard coded for now

justinmc committed:
Animation fully reverses itself at the right time

justinmc committed:
Mask works

justinmc committed:
Cleanup extra widget

justinmc committed:
Fix analyze issues

justinmc committed:
Open contextmenu layout for child and menu

justinmc committed:
Properly lined up animations, and correct final positions

justinmc committed:
Menu looks and behaves like a menu.

justinmc committed:
Tap event on the child itself exposed to consumer

justinmc committed:
Fit all sized children into all screen sizes.

justinmc committed:
Horizontal padding

justinmc committed:
Menu action border, and color tweaks

justinmc committed:
License header for new test file

justinmc committed:
Support regular taps on child via GestureDetector. Animations don't interfere.

justinmc committed:
Clean up animation code

justinmc committed:
Fix menu appearing under cutout when landscape

justinmc committed:
Row/Column layout based on orientation

justinmc committed:
Optional icon and default/destructive styles

justinmc committed:
WIP Experiment with using a Hero approach

justinmc committed:
Revert "WIP Experiment with using a Hero approach"
I'm going to explore doing it without a hero for now but want to keep
the history.

This reverts commit 408159686d67bcac15dda52d2bb0609c1dfac432.

justinmc committed:
Unanimated, normal positioning. To be animated later with another dummy.

justinmc committed:
Render one frame offstage and measure final position

justinmc committed:
Do initial bounce animation in Overlay to avoid z-index problems

justinmc committed:
Perfect sizing for modal transition and final position

justinmc committed:
Rerunnable

justinmc committed:
Close animation lines up perfectly with original

justinmc committed:
ContextMenuSheet animates as well

justinmc committed:
Fix bug where rotating the phone when contextmenu open causes reverse animation to go to old position

justinmc committed:
Simmplify isopening and isopen

justinmc committed:
Fix analyze problems

justinmc committed:
Fix misused scale for the initial bounce animation

justinmc committed:
Proper canceling of open when tap isnt held long enough

justinmc committed:
Reverse animations don't show mask color

justinmc committed:
Drag down to dismiss ContextMenu

justinmc committed:
Fling

justinmc committed:
Scale with drag

kerberjg commented:
This looks really promising! Thank you for your contribution, can't wait to see this LGTM'd :)

justinmc committed:
Animate the scale

justinmc committed:
Vibrate on open

justinmc committed:
Make dismiss fling super damped like in updated ios13 beta

justinmc committed:
Fix bug where tapping on bg during animation threw error. Now will reverse animate with intermediate positions

justinmc committed:
Fix bug where child didn't line up at open time and close time

justinmc committed:
More close match for vibration

justinmc committed:
Allow for ellipsizing of overflowing actions

justinmc committed:
Fix error when flinging towards start

justinmc committed:
ContextMenuSheet disappears when dismissing

justinmc committed:
More native-like dismiss animation for contextmenusheet

justinmc committed:
Align child properly in landscape

justinmc committed:
Adapt order of ContextMenuSheet and child based on which side of the screen the original child is location in

justinmc committed:
Fix bug where child was visible while menu was open

justinmc committed:
Sheet animation is in the right place in all orientations and places on screen

justinmc committed:
ContextMenuSheet scales in right direction when being dragged

justinmc committed:
Drag to dismiss when in landscape as well

justinmc committed:
Fix bug where releasing drag doesn't scale back up, and hide Sheet a little before minimum scale

justinmc committed:
Damped negative vertical dragging

justinmc committed:
Reverse drag scale increase

justinmc committed:
preview param

justinmc committed:
dummy => decoy

justinmc committed:
Simplify openContextMenu

justinmc committed:
Clean up some TODOs

justinmc committed:
Move ContextMenuSheetAction to its own file to slim down context_menu.dart

justinmc committed:
Clean up TODOs and analyze errors

justinmc committed:
Basic open and close tests

justinmc committed:
Fling tests

justinmc committed:
ContextMenuSheetAction test

justinmc committed:
Fix analyze errors

justinmc committed:
Bug fixed where, after fling, animation back to home didn't happen

justinmc commented:
@HansMuller @LongCatIsLooong Thanks in advance for the review!  Some things up front:

  * Let me know what you think of the overall API for using this sort of widget.
  * I left a bunch of TODOs in the code where I wanted to start a discussion with you guys, so please comment when you see those.
  * This could still use another pass to get the animations and sizing closer pixel perfect with iOS.
  * @LongCatIsLooong I have some hardcoded light/dark theme stuff, but now that you've start implementing CupertinoDynamicColor, can I go ahead and plug that in here?

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1099 @@
+
+  // The OverlayEntry that positions widget.child directly on top of its
+  // original position in this widget.
+  OverlayEntry get _overlayEntry {
```
It feels a bit weird to have a getter that returns a new OverlayEntry every time it gets accessed.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1099 @@
+  @override
+  Widget build(BuildContext context) {
+    return GestureDetector(
+      onTapCancel: _onTapCancel,
```
Is it possible to use a `LongPressGestureRecognizer`, instead of using `isAnimating` to tell if the gesture should be accepted as a long press?

justinmc replied:
Compromise: I changed this to base the whole animation on the duration of a long press, which I think is a lot more robust than my original approach.  I didn't directly use `onLongPress` though.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1099 @@
+    // Also, I noticed that I'm having the same theme problem where a widget in
+    // the Overlay might not be themed the same as it is otherwise. Sizing can
+    // also potentially be a problem if the child was sized by its parent before.
+    return OverlayEntry(
```
Is it possible to avoid using the additional `OverlayEntry`? So we don't have to manage it and thus avoid the `Future.delayed` hack. Maybe make the expand animation part of the `ContextMenuRoute` push transition, so that `_kOpenScale` will be <= 1.0?

justinmc replied:
I experimented with a few different ways to do this.  For not using an Overlay, it seems to come down to the fact that the child will appear behind its siblings during its bounce animation.  This looked really obvious to me when I tried it.  I updated the comment here to be a bit more clear about the situation at least.

LongCatIsLooong commented on packages/flutter/test/cupertino/context_menu_test.dart:
```diff
@@ -0,0 +1,195 @@
+  Finder _findDecoyChild(Widget child) {
+    return find.descendant(
+      of: find.byType(ShaderMask),
+      matching: find.byWidgetPredicate((Widget w) => w == child),
```
I think this is equivalent to `find.byWidget(child)`?

justinmc replied:
You appear to be right, thanks for the tip.

HansMuller reviewed (commented):
Some feedback that I'd forgotten to deliver a week or two ago...

justinmc committed:
Merge branch 'master' into ios13-context-menus

justinmc committed:
Simple code review fixes, docs, etc.

justinmc committed:
Give up on fixing the 1 frame offscreen for now after investigating some alternatives

justinmc committed:
Clean up todos

justinmc committed:
Split bounce into 2 animations and work on pixel perfectness

justinmc committed:
Test alignment of actions

justinmc committed:
Fix analyze errors

justinmc committed:
Base decoy animation on longpress and prevent stopping the open after the vibration

justinmc committed:
Comment improvements

justinmc committed:
Tool examples, and a change to make it work with normal Containers (no FittedBox, user should pass it in if needed)

justinmc committed:
Eliminate OverlayEntry getter

justinmc committed:
Code cleanup

justinmc commented:
@HansMuller @LongCatIsLooong I was finally able to come back to this and get it ready for re-review.  Let me know what you think.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1222 @@
+  ///       ],
+  ///     ),
+  ///   ],
+  /// );
```
From the looks of it, this semicolon might be what made the analyzer fail

justinmc replied:
I will try pushing that and see if it works.  Thank you!  I would never have thought of that one.

justinmc replied:
That was it!

justinmc committed:
Try removing a semicolon in comments for the analyzer

HansMuller reviewed (commented):
Just a progress report. Don't understand the decoy stuff yet.

justinmc commented:
@LongCatIsLooong pointed out that native supports submenus.  I've created an issue to add this feature later if that's ok with the reviewers. https://github.com/flutter/flutter/pull/37778

justinmc commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1222 @@
+/// Widget build(BuildContext context) {
+///   return Scaffold(
+///     key: scaffoldKey,
+///     body: Center(
```
Do you mean that I can get rid of the Center widget and just use `alignment: Alignment.center`?  It didn't seem to work for me, the red square ended up in the top left when I did this:

```dart
      body: Container(
        width: 100,
        height: 100,
        alignment: Alignment.center,
        child: ContextMenu(
...
```

Or maybe you just mean that I should add `alignment: Alignment.center` to the Container?

HansMuller replied:
My mistake, what you had is fine.

justinmc commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1222 @@
+
+// The context menu arranges itself slightly differently based on the location
+// on the screen of [ContextMenu.child] before the [ContextMenu] opens.
+enum _ContextMenuOrientation {
```
Yeah that seems more accurate to me :+1:

justinmc commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1099 @@
+/// Long pressing or 3d touching on the child will open in up in a full-screen
+/// overlay menu.
+class ContextMenu extends StatefulWidget {
+  /// Create a context menu.
```
Good to know, thanks.

justinmc commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1222 @@
+  ///
+  /// This parameter cannot be null, and items in the List must be
+  /// [ContextMenuSheetAction]s.
+  final List<ContextMenuSheetAction> actions;
```
I was following the pattern of CupertinoActionSheetAction, which is probably not a great name to follow, so I'm on board with changing it.

justinmc committed:
Review docs fixes and ContextMenuOrientation name change

justinmc committed:
ContextMenuSheetAction => ContextMenuAction

justinmc commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1222 @@
+  /// [ContextMenuSheetAction]s.
+  final List<ContextMenuSheetAction> actions;
+
+  /// The callback to call when tapping on the child when the [ContextMenu] is
```
Now that I look into this, it seems to work when both open and closed by just wrapping the child in a GestureDetector.

Maybe the real solution is to provide a way for the user to know when the ContextMenu is open vs. closed?  For now I think I'm just going to remove this callback.

justinmc committed:
Remove onTap callback, wrap child in GestureDetector instead

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1224 @@
+class ContextMenu extends StatefulWidget {
+  /// Create a context menu.
+  ///
+  /// [actions] is required and cannot be null. Itmust contain at least one
```
```suggestion
  /// [actions] is required and cannot be null. It must contain at least one
```

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1224 @@
+/// See also:
+///
+///   * [Apple's HIG for Context Menus](https://developer.apple.com/design/human-interface-guidelines/ios/controls/context-menus/)
+class ContextMenu extends StatefulWidget {
```
Do we want the `Cupertino` prefix? I found this PR https://github.com/flutter/flutter/pull/20267 but couldn't find the rationale.

justinmc replied:
CC @HansMuller

HansMuller replied:
I think we've established a convention already:
```
CupertinoActivityIndicator
CupertinoDatePicker
CupertinoTimerPicker
CupertinoTabBar
CupertinoTabView
CupertinoTabScaffold
CupertinoSegmentedControl<T>
CupertinoSliverRefreshControl
CupertinoButton
CupertinoUserInterfaceLevel
CupertinoPicker
CupertinoApp
CupertinoSlider
CupertinoPageScaffold
CupertinoAlertDialog
CupertinoDialog
CupertinoPopupSurface
CupertinoDialogAction
CupertinoTheme
CupertinoTextField
CupertinoNavigationBar
CupertinoSliverNavigationBar
CupertinoNavigationBarBackButton
CupertinoSwitch
CupertinoActionSheet
CupertinoActionSheetAction
CupertinoScrollbar
CupertinoTextSelectionToolbar
```

justinmc replied:
That looks like a trend to me haha.  I'm changing it.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1224 @@
+
+/// A full-screen modal route that opens when the [child] is long-pressed.
+///
+/// When open, the ContextMenu shows the child, or [preview] if given, in a
```
nit: `ContextMenu` can be put in square brackets (or should it?), so are `Overlay`, `Expanded` and `PopupRoute`.

justinmc replied:
Done, but with a caveat.  I left ContextMenu unbracketed because this is the definition of ContextMenu itself, so the link would go to the same page.  Is that correct, or should I still bracket it?

LongCatIsLooong replied:
I asked the same question in different PR. @goderbauer's answer was we should still square bracket it.

justinmc replied:
Alright I'm bracketing that too.

Also, should I only bracket the first occurrence?  That's what I've been doing to avoid having lots of duplicate links right next to each other.

LongCatIsLooong commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1224 @@
+///
+/// The [preview] parameter is most commonly used to display a slight variation
+/// of [child]. See [preview] for an example of rounding the child's corners and
+/// allowing its aspect ratio to expand, similar to the iPhoto app on iOS.
```
nit: I think the app is just called "Photos".

justinmc replied:
Apparently iPhoto is the Mac desktop photo editing app discontinued in 2015, whoops :) Good catch.

justinmc commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1222 @@
+  @override
+  void initState() {
+    super.initState();
+    _decoyController = AnimationController(
```
I agree, I changed it to `_openController` and added a bit of explanation as well.

justinmc committed:
Code review docs and naming improvements

justinmc committed:
More code review docs improvements

justinmc commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1222 @@
+  ///         // ContextMenu is open, even though it's not rounded when it's
+  ///         // closed.
+  ///         child: ClipRRect(
+  ///           borderRadius: BorderRadius.circular(64.0),
```
That's a great idea that I should have thought of.  I'm looking into it.

justinmc committed:
preview => previewBuilder

justinmc committed:
Use addPostFrameCallback instead of Future.delayed, to be more explicit

justinmc committed:
ContextMenu => CupertinoContextMenu, ContextMenuAction => CupertinoContextMenuAction

justinmc commented:
@HansMuller @LongCatIsLooong Ready for re-review.  I've renamed to CupertinoContextMenu, changed `preview` to `previewBuilder`, and made a bunch of smaller requested changes.

One question I had: Is there a callback type that I can reuse instead of defining the PreviewBuilder and _PreviewBuilderChildless types ?

justinmc committed:
Bracket link to itself

HansMuller commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1274 @@
+  /// The widget that can be "opened" with the [CupertinoContextMenu].
+  ///
+  /// When the [CupertinoContextMenu] is long-pressed, the menu will open and
+  /// this widget (or the widget returned by [previewBuilder], if provided) will
```
I think this may be revealing a detail that's not important: that the child/previewBuilder will be moved to the overlay. Can we explain what it means to "open" the menu just in terms of the route we're going to push? That approach jibes with using Navigator.pop() to close the menu.

HansMuller commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1274 @@
+
+  /// The actions that are shown in the menu.
+  ///
+  /// This parameter cannot be null, and items in the List must be
```
can't be null or empty

HansMuller commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1274 @@
+  /// sharp corners when in the page.
+  ///
+  /// In addition to the current [BuildContext], the function is also called
+  /// with an [Animation] and the [child]. The animation is used internally to
```
instead of "is used internally", you could explain that the animation goes from 0-1 when the menu opens, 1-0 when it closes.

HansMuller commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1274 @@
+  ///       // it's closed. It uses the given animation to animate the corners
+  ///       // in sync with the opening animation.
+  ///       child: ClipRRect(
+  ///         borderRadius: BorderRadius.circular(64.0 * animation.value),
```
NICE

HansMuller commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1099 @@
+  // child in the screen.
+  _ContextMenuOrientation get _contextMenuOrientation {
+    final Rect childRect = _getRect(_childGlobalKey);
+    final double screenWidth = MediaQuery.of(context).size.width;
```
What's the answer here?

justinmc replied:
Sorry, I swear I responded to this before but I can't find my comment.  This code is used to find the location in the overall screen of the original file.  The reason is that in the open CupertinoContextMenu, the actions are displayed either left/center/right aligned horizontally based on this position of the child.  So, it's necessary to use the full screen size here.

I'm adding a comment to this getter to explain that.

HansMuller commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1274 @@
+
+// A floating copy of the CupertinoContextMenu's child.
+//
+// When the child is pressed, but before the CupertinoContextMenu opens, it does
```
Thanks for the explanation!

HansMuller commented on packages/flutter/lib/src/widgets/routes.dart:
```diff
@@ -701,7 +702,11 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
+  }) : _filter = filter,
+       super(settings: settings);
+
+  final ui.ImageFilter _filter;
```
Doc needed

This new route feature should also get its own test.

HansMuller reviewed (commented):
This looks good; just some small feedback plus: need to document/test ModalRoute.filter

justinmc committed:
Code review, docs and test for filter, fix test naming

justinmc committed:
Fix naming in CupertinoContextMenuAction test

justinmc commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1274 @@
+      flex: 2,
+      child: IntrinsicHeight(
+        child: ClipRRect(
+          borderRadius: BorderRadius.circular(16.0),
```
Thank you!

justinmc commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1274 @@
+  ///
+  /// This parameter cannot be null, and items in the List must be
+  /// [CupertinoContextMenuAction]s.
+  final List<CupertinoContextMenuAction> actions;
```
Yeah there is nothing preventing us from using `List<Widget>` here.  I did it just to encourage users to always use CupertinoContextMenuActions, but I also didn't realize there were group separators.  Does Flutter usually allow any type of widget when it can?

LongCatIsLooong replied:
There's an issue for this: https://github.com/flutter/flutter/issues/12078. Also I tried adding a context menu to a `UISegmentedControl`, and it turns out other than group separators there can also be headers (similar to action sheet headers)

justinmc replied:
If that's more in line with how we usually do it then changing it sounds good to me.

justinmc replied:
Oh sorry I posted my last comment before Github updated to show yours.  That issue definitely confirms that this should be agnostic so I'm changing it now :+1:

justinmc commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1274 @@
+  /// Below is an example of using `previewBuilder` to show an image tile that's
+  /// similar to each tile in the iOS iPhoto app's context menu. Several of
+  /// these could be used in a GridView for a similar effect.
+
```
Good catch

justinmc committed:
Code review docs fix and ios border radius value

justinmc commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1274 @@
+
+class _CupertinoContextMenuState extends State<CupertinoContextMenu> with TickerProviderStateMixin {
+  final GlobalKey _childGlobalKey = GlobalKey();
+  double _childOpacity = 1.0;
```
I'm trying this out.

Actually @goderbauer I got this code from you awhile back.  I was asking about animating to a final laid out state, and you suggested rendering the first frame offstage and sent me this gist: https://gist.github.com/goderbauer/b136e871fac33960dbd539ea1930ff97

Any idea if there is an advantage to Offstage and TickerMode instead?

justinmc replied:
Ah I think it's because PopupRoute is already using Offstage internally, and so this widget is just managing the `offstage` variable.

https://github.com/flutter/flutter/blob/ed36fe5ddde84ffd9b6380687bc6a2f37b0790ab/packages/flutter/lib/src/widgets/routes.dart#L663

LongCatIsLooong replied:
What I originally meant was to use `Offstage` + `TickerMode` to hide the original child instead of `Opacity`. Should we stop the animation in the original child when it's hidden, by adding a `TickerMode` parent? Toggling `Opacity` between 0 and 1 seems fine.

justinmc replied:
Oh I see, sorry!  That makes more sense and is probably more performant.  I'll change it.

justinmc replied:
Actually after changing it something broke.  I was getting the size of the Opacity widget using a GlobalKey, and I couldn't do that with an Offstage widget (even when it was onstage).  It gave me weird values and broke the tests.

My compromise is to use a TickerMode and an Opacity widget together, but let me know if you know why getting the size wasn't working.

justinmc committed:
List<CupertinoContextMenuAction> => List<Widget>

justinmc committed:
Remove unused import

justinmc committed:
kOpenScale was measured to be 1.1 in Xcode.

HansMuller commented on packages/flutter/lib/src/cupertino/context_menu.dart:
```diff
@@ -0,0 +1,1278 @@
+  final Widget child;
+
+  /// The actions that are shown in the menu.
+  ///
```
This would be a good place to say that typically the actions are [CupertinoContextMenuAction]s.

HansMuller reviewed (approved):
LGTM

justinmc committed:
Docs addition

justinmc committed:
Update kopenscale in tests too, whoops

justinmc committed:
Merge branch 'master' into ios13-context-menus

justinmc committed:
Merge branch 'master' into ios13-context-menus

justinmc committed:
Use Offstage and TickerMode instead of Opacity

justinmc committed:
Merge branch 'master' into ios13-context-menus

justinmc committed:
Merge branch 'master' into ios13-context-menus

justinmc commented:
Closing in favor of: https://github.com/flutter/flutter/pull/43918
